---
title: "sims"
output: html_document
---
 
```{r}
library(wingen)
library(raster)
library(vcfR)
library(viridis)
library(foreach)
library(doParallel)
library(here)
library(ggplot2)
library(dplyr)
library(purrr)

library(adegenet)
library(hierfstat)
source("sims_functions.R")
```


# system time sims
```{r}
vcf <- read.vcfR("data/mod-sim_params_it-0_t-8000_spp-spp_0.vcf")
geo <- read.csv("data/mod-sim_params_it-0_t-8000_spp-spp_0.csv")
coords <- geo[,c("idx","x","y")]

set.seed(22)
l <- sample(nrow(vcf@gt), 10000)
si <- sample(nrow(coords), 200)
#write.csv( data.frame(loci = l, inds = si), "samples_seed42.csv", row.names = FALSE)

subvcf <- vcf[l,c(1, si+1)]

subcoords <- coords[si,]
subcoords <- subcoords[,c("x","y")]
subcoords$y <- -subcoords$y
coords$y <- -coords$y

# check match
all(colnames(subvcf@gt)[-1] == geo$idx[si])

```

```{r,  fig.width = 6, fig.height = 6}
lyr <- read.csv("data/middle_earth.csv", header = FALSE)
lyr <- raster(as.matrix(lyr))
extent(lyr) <- extent(0,100,-100,0)

kde <- raster(MASS::kde2d(coords$x, coords$y, h = c(10,10), n = 100, lims = c(0,100,-100,0)))

par(mar = rep(2,4))
plot(lyr, col = viridis::magma(100), main = "Carrying Capacity", axes = FALSE, box = FALSE)
plot(kde, col = viridis::magma(100), main = "Population Density", axes = FALSE, box = FALSE)
plot(lyr, col = mako(1, begin = 0.1, alpha = 0.1), main = "Distribution of Individuals", axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
points(coords$x, coords$y, pch = 16, cex = 1, col = mako(1, begin = 0.7, alpha = 0.4), xlab = "", ylab = "")
points(subcoords$x, subcoords$y, pch = 16, cex = 1, col = mako(1, begin = 0.2, alpha = 0.9), xlab = "", ylab = "")
legend(0,-80, 
       c("All Individuals", "Sampled Individuals"), 
       pch = c(16,16), 
       col = c(mako(1, begin = 0.7), mako(1, begin = 0.2)),
       bty = "n",
       text.col = "black",
       cex = 1.5)

```


```{r, fig.width = 6, fig.height = 6, cache = TRUE}
preview_gd(lyr, subcoords, wdim = 5, fact = 3)
cores <- 2
cl <- makeCluster(cores)
registerDoParallel(cl)

tr <- sim(subvcf, subcoords, lyr, stat = "het", wdim = 5, fact = 3, rarify_n = 4, rarify_nit = 5, rarify = TRUE, parallel = TRUE, nloci = nrow(subvcf@gt))
stopCluster(cl)

tr_k <- krig_gd(tr[["res"]], lyr, disagg_grd = 3)
tr_k_mask <- mask_gd(tr_k, 4)

par(mar = rep(0,4), oma = rep(0,4), pty = "s")
plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
plot(tr[["res"]][[1]], col = viridis::magma(100, alpha = 1.0), main = "Window Pi", axes = FALSE, box = TRUE, legend = TRUE, add = TRUE)

plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
plot(tr_k[[1]], col = viridis::magma(100), main = "Masked Pi", axes = FALSE, box = TRUE, legend = FALSE, add = TRUE)

plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
plot(tr_k_mask, col = viridis::magma(100), main = "Kriged and Masked Pi", axes = FALSE, box = TRUE, legend = FALSE, add = TRUE)

par(mar = rep(0,4), oma = rep(0,4), pty = "s")
plot(tr[["res"]]$sample_count, col = viridis::mako(100, alpha = 1.0), axes = FALSE, box = TRUE, legend = TRUE)
plot(tr_k$sample_count, col = viridis::mako(100), axes = FALSE, box = TRUE, legend = FALSE)

```


```{r, fig.width = 16, fig.height = 4}
par(mfrow = c(1,4), mar = rep(1,4), oma = rep(0,4), pty = "s")

tr_k_2 <- krig_gd(tr[["res"]], lyr, disagg_grd = 2, resample = TRUE)
for(i in c(1, 2, 4, 8)){
  tr_k_mask <- mask_gd(tr_k_2, i)
  plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
  plot(tr_k_mask, col = viridis::magma(100), axes = FALSE, box = TRUE, legend = FALSE, add = TRUE, zlim = c(0.18, 0.47))
}

```

```{r, fig.width = 16, fig.height = 4}
par(mfrow = c(1,4), mar = rep(1,4), oma = rep(0,4), pty = "s")

tr_k_2F <- krig_gd(tr[["res"]], lyr, disagg_grd = 2, resample = FALSE)
for(i in c(1, 2, 4, 8)){
  tr_k_mask <- mask_gd(tr_k_2F, i)
  plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
  plot(tr_k_mask, col = viridis::magma(100), axes = FALSE, box = TRUE, legend = FALSE, add = TRUE, zlim = c(0.18, 0.47))
}

```
```{r, fig.width = 16, fig.height = 4}
par(mfrow = c(1,4), mar = rep(1,4), oma = rep(0,4), pty = "s")

tr_k_2F <- krig_gd(tr[["res"]], lyr, disagg_grd = 1, resample = FALSE)
for(i in c(1, 2, 4, 8)){
  tr_k_mask <- mask_gd(tr_k_2F, i)
  plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
  plot(tr_k_mask, col = viridis::magma(100), axes = FALSE, box = TRUE, legend = FALSE, add = TRUE, zlim = c(0.18, 0.47))
}

```
```{r, fig.width = 16, fig.height = 4}
par(mfrow = c(1,4), mar = rep(1,4), oma = rep(0,4), pty = "s")

tr_k_2F <- krig_gd(tr[["res"]], lyr, disagg_grd = 3, resample = FALSE)
for(i in c(1, 2, 4, 8)){
  tr_k_mask <- mask_gd(tr_k_2F, i)
  plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
  plot(tr_k_mask, col = viridis::magma(100), axes = FALSE, box = TRUE, legend = FALSE, add = TRUE, zlim = c(0.18, 0.47))
}

```
# Comparison of measures

# Nsamp = 100
```{r, fig.width = 6, fig.height = 8}

measures <- c("pi", "biallelic_richness", "heterozygosity")
datasets <- c("rr", "WGS", "FULL")
for(nsamp in c(100, 200)){
  msk_lyr <- get_divout(file.name = "rr", rarify = TRUE, measure = "pi", nsamp = nsamp)
  for(m in measures){
    par(mfrow = c(3, 2), mar = rep(2,4))
    for(d in datasets){ 
      for(rarify in c("TRUE", "FALSE")){
        r <- get_divout(file.name = d, rarify = rarify, measure = m, nsamp = nsamp)
        r <- mask(r, msk_lyr)
        plot(r,  col = magma(100), box = FALSE, axes = FALSE, main = paste(d, m, rarify))          
      }
    }
  }
}



```

```{r, fig.width = 7.5, fig.height = 5}
# {PLOT CODE}
for(nsamp in c(100,200)){
  msk_lyr <- get_divout(file.name = "rr", rarify = TRUE, measure = "pi", nsamp = nsamp)
  for(m in measures){
    par(mfrow = c(2, 3), mar = c(1,0,1,0), oma = rep(0,4))
    for(rarify in c("TRUE", "FALSE")){
      for(d in datasets){
        r <- get_divout(file.name = d, rarify  = rarify, measure = m, nsamp = nsamp)
        r <- mask(r, msk_lyr)
        if(m == "pi"){zlim = c(0.1,0.5)}
        if(m == "biallelic_richness"){zlim = c(1.1,2)}
        if(m == "heterozygosity"){zlim = c(0,1)}
         plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = TRUE, zlim = c(0.01,1), legend = FALSE)
         plot(r,  col = magma(100), box = TRUE, axes = FALSE, main = "", zlim = zlim, legend = FALSE, add = TRUE)          
      }
    }
  }
}

```


```{r}
# demonstrate that allelic richness and biallelic richness are the same 
wgs_ar <- stack("outputs/WGS_200ind_allelic_richness.tif")[[1]]

wgs_bar <- stack("outputs/WGS_200ind_biallelic_richness.tif")[[1]]

par(mfrow = c(1,2))
lapply(list(wgs_bar, wgs_ar), raster::plot, col = magma(100), box = FALSE, axes = FALSE)

```

## Timing
```{r}


tdf <- dplyr::bind_rows(
  map_dfr(c("rr", "WGS"), get_timeout, rarify = "TRUE", parallel = "FALSE", nsamp = 100),
  map_dfr(c("rr", "WGS"), get_timeout, rarify = "TRUE", parallel = "TRUE", nsamp = 100),
  
  map_dfr(c("rr", "WGS"), get_timeout, rarify = "TRUE", parallel = "FALSE", nsamp = 200),
  map_dfr(c("rr", "WGS"), get_timeout, rarify = "TRUE", parallel = "TRUE", nsamp = 200)
)

tdf[tdf$dataset == "rr", "dataset"] <- "RADseq-type dataset (10,000 loci)"
tdf[tdf$dataset == "WGS", "dataset"] <- "WGS-type dataset (100,000 loci)"

ggplot(data = tdf) +
  geom_hline(yintercept = 60, linetype = "dashed", col = "darkgray", lwd = 1) + 
  geom_text(aes(factor(200), 60, label = "1 min", vjust = -1, hjust = -0.7), col = "darkgray") +
  geom_hline(yintercept = 60*5, linetype = "dashed", col = "gray", lwd = 1) + 
  geom_text(aes(factor(200), 60*5, label = "5 min", vjust = -1, hjust = -0.7), col = "gray") +
  geom_bar(aes(x = factor(nsamp), y = time, fill = stat), 
           stat = "identity", 
           position=position_dodge()) +
  facet_grid(~dataset,  scales = "free_y") +
  scale_fill_viridis_d(option = "mako", begin = 0.3, end = 0.8, name = "statistic") +
  xlab("number of samples") +
  ylab("time (seconds)") +
  theme_bw() +
  theme(panel.grid.minor.y = element_blank(), 
        panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank(), 
        panel.grid.major.x = element_blank())

```

```{r}
load("RData/ar_results.RData")

```
## Window vs aggregation factor

```{r, fig.width = 5, fig.height = 5}
cores <- 6
cl <- makeCluster(cores)
registerDoParallel(cl)

stk <- list()

i = 1
for(f in c(2, 4, 8)){
  for(w in c(3, 5, 7)){
    res <- window_gd(subvcf, subcoords, lyr, stat = "pi", wdim = w, fact = f, rarify = TRUE, rarify_n = 4, rarify_nit = 5, parallel = TRUE, nloci = nrow(subvcf@gt))
    stk[[i]] <-  res[[1]]
    plot(res[[1]], col = magma(100))
    i <- i + 1
  }
}

i = 1
par(mfrow = c(3,3), mar = rep(0,4), oma = rep(0,4), pty = "s")
for(f in c(2, 4, 8)){
  for(w in c(3, 5, 7)){
    plot(lyr, col = magma(1, begin = 0.1, alpha = 0.1), axes = FALSE, box = FALSE, zlim = c(0.01,1), legend = FALSE)
    plot(stk[[i]], col = magma(100), axes = FALSE, box = FALSE, legend = FALSE, zlim = c(0.2, 0.5), add = TRUE)
    rect(0, -105, 100, 0, border = "black")
    i <- i + 1
  }
}
plot(stk[[1]], zlim = c(0.23, 0.48), col = magma(100) )
```

#
# IGNORE BELOW THIS LINE - RAREFACTION STUFF --------------------------------------------------------

```{r, fig.width = 4, fig.height = 4}
set.seed(42)

resf <- purrr::map(c(2, 4, 8, 16), time_test, "fact", subvcf, subcoords, lyr, stat = "pi", wdim = 5, rarify = TRUE, rarify_n = 10, rarify_nit = 10)

plot_time_test(resf, stat = "fact")
plot_time_test(resf, stat = "ncell")
plot_time_test(resf, stat = "total_count")
```

```{r, fig.width = 10, fig.height = 5}
plot_rast_test(resf)
```

```{r, fig.width = 16, fig.height = 4}
set.seed(42)

resw <- purrr::map(c(3, 5, 7, 9, 11), time_test, "wdim", subvcf, subcoords, lyr, fact = 3, stat = "pi", rarify = TRUE, rarify_n = 4, rarify_nit = 5)

plot_time_test(resw)

resw <- purrr::map(c(3, 5, 7, 9, 11), time_test, "wdim", subvcf, subcoords, lyr, fact = 5, stat = "pi", rarify = TRUE, rarify_n = 4, rarify_nit = 5)

plot_time_test(resw)

```

```{r, fig.width = 10, fig.height = 4}
plot_rast_test(resw)
```

```{r, fig.width = 16, fig.height = 4}
set.seed(42)

resfr <- purrr::map(c(2, 4, 8, 16), time_test, "fact", subvcf, subcoords, lyr, stat = "pi", wdim = 5, rarify = FALSE, min_n = 10)

plot_time_test(resfr)

```

```{r, fig.width = 10, fig.height = 4}
plot_rast_test(resfr)
```

```{r, fig.width = 16, fig.height = 4}
set.seed(42)

reswr <- purrr::map(c(5, 7, 9, 11, 13), time_test, "wdim", subvcf, subcoords, lyr, fact = 4, stat = "pi", rarify = FALSE, min_n = 10)

plot_time_test(reswr)

```
# REALIZATION: TIME CHANGING MIGHT HAVE TO DO WITH EDGE OF LANDSCAPE, NOT WINDOW SIZE EXACLTY - MAY NEED TO TRY FAKE LANDSCAPE WITH PERFECT EVEN SAMPLING TO GET CALCS RIGHT
(replicate the vcf from one individual multiple times evenly across a landscape, also check if there is a diff based on the value but there shouldn't be)

# UNIFORM GRID
```{r, fig.width = 5, fig.height=5}
m <- matrix(1, ncol = 10, nrow = 10)
r <- raster(m)
pts <- rasterToPoints(r)
dcoords <- data.frame(pts[,c("x","y")])
dvcf <- vcf[1:10000,c(1, sample(nrow(dcoords)) + 1)]

dr <- raster(matrix(1,ncol = 10, nrow = 10))

par(pty ="s")
ncell(dr) == nrow(dcoords)
plot(dr)
points(dcoords)
```
note: rarify_n needs to be 2 so that landscape/sampling doesn't affect the results (e.g. you don't want any NA values)
```{r}
dr100 <- raster(matrix(1, ncol = 100, nrow = 100))
resf3 <- purrr::map(c(2, 3, 5, 10, 20), time_test, "fact", dvcf, dcoords, dr100, stat = "pi", wdim = 3, rarify = FALSE, min_n = 1)

plot_time_test(resf5, stat = "fact")
```
# FIGURE OUT THIS WEIRDNESS:
```{r}
plot(dr)
plot(crop(dr, extent(0,21/100,0,21/100)), add = TRUE, col = "red")
points(dcoords)

resf <- purrr::map(c(3, 5, 7, 9), time_test, "wdim", dvcf, dcoords, dr, stat = "pi", rarify = FALSE, min_n = 1)
plot_time_test(resf, stat = "wsize")
plot_time_test(resf)

plot_rast_test(resf)
```

note to self: I don't think I am counting total time right still...need to think on it (check the minutes versus seconds calculations)


# Rarefaction test

```{r}
m <- matrix(1, ncol = 10, nrow = 20)
r1 <- raster(m)
extent(r1) <- c(0, 10, 0, 20)
m <- matrix(2, ncol = 10, nrow = 20)
r2 <- raster(m)
r2 <- aggregate(r2, 2)
extent(r2) <- c(10, 20, 0, 20)
pts <- rbind(rasterToPoints(r1), rasterToPoints(r2))


dcoords <- data.frame(pts[,c("x","y")])
plot(dcoords)
dvcf <- vcf[1:10000,c(1, sample(nrow(dcoords)) + 1)]

dr <- raster(matrix(1, ncol = 20, nrow = 20))
extent(dr) <- c(0,20,0,20)
par(pty ="s")
plot(dr)
points(dcoords)

wr <- sim(dvcf, dcoords, dr, stat = "het", wdim = 7, fact = 0, min_n = 6, rarify = FALSE, parallel = TRUE)

wor <- sim(dvcf, dcoords, dr, stat = "het", wdim = 7, fact = 0, rarify = TRUE, rarify_n = 6, rarify_nit = 10, parallel = TRUE)

plot(wr$res[[1]], zlim = c(4029, 4829.786), col = magma(100))
plot(wor$res[[1]], zlim = c(4029, 4829.786), col = magma(100))

par(mfrow = c(1,2))
plot(wr$res[[1]], zlim = c(0, 0.5), col = magma(100), main = "No Rarefaction")
plot(wor$res[[1]], zlim = c(0, 0.5), col = magma(100), main = "Rarefaction")
```



#Rarefaction take 2
```{r}
SiteSampleBuffer <- function(sample_sites, coords, npts, buffer_size = 5){
  #sample_sites - coordinates of sampling sites
  #coords - coordinates of all individuals in data set
  #npts - number of points to sample from each site
  #buffer - buffer around site from which to draw samples randomly
  #edge_buffer - buffer from landscape edges to prevent sampling of sites
  
  site_samples <- data.frame()
  
  #sample sites can be provided as either sp or coords in vector format (nloop just counts how many sites)
  if(class(sample_sites)[1] == "SpatialPoints"){nloop <- length(sample_sites)} else {nloop <- nrow(sample_sites)}
  
  for(s in 1:nloop){
    #create buffer around sites from which to sample points
    site_buffers <- rgeos::gBuffer(sample_sites[s,], width = buffer_size)
    #subset out only coordinates falling within site buffers
    buffer_samples <- coords[site_buffers,]
    if(length(buffer_samples) < npts){stop(paste0("Less than npts found in buffer around site"))}
    #convert from SPDF to df
    buffer_samples_df <- data.frame(buffer_samples)
    #randomly sample samples from coordinates
    randsamp <- sample(nrow(buffer_samples_df), npts)
    #subset df
    buffer_samples_df <- buffer_samples_df[randsamp,]
    
    #save site IDs
    buffer_samples_df$site <- s
    
    #bind samples
    site_samples <- rbind(site_samples, buffer_samples_df)
  }
  return(site_samples[,c("idx", "x", "y")])
}
```

```{r}

plot(lyr)
points(coords)

pts <- coords
coordinates(pts) <- ~x+y

sample_sites <- rbind.data.frame(c(75, -20), c(25, -20), c(50, -20))
colnames(sample_sites) <- c("x", "y")
coordinates(sample_sites) <- ~x+y


subcoords <- rbind.data.frame(SiteSampleBuffer(sample_sites[3], pts, npts = 5, buffer_size = 10),
                              SiteSampleBuffer(sample_sites[2], pts, npts = 20, buffer_size = 10))


plot(lyr)
points(subcoords[,c("x","y")])


si <- colnames(vcf@gt) %in% subcoords$idx
si[1] <- TRUE
subvcf <- vcf[l, si]

# reorder coords to match
subcoords <- subcoords[match(colnames(subvcf@gt[,-1]), subcoords$idx),]
 
# check match
all(colnames(subvcf@gt)[-1] == as.character(subcoords$idx))

nr <- window_gd(subvcf, subcoords[,c("x","y")], lyr, "biallelic.richness", wdim = 5, fact = 3, rarify = FALSE, min_n = 4, nloci = nrow(subvcf@gt))
yr <- window_gd(subvcf, subcoords[,c("x","y")], lyr, "biallelic.richness", wdim = 5, fact = 3,rarify = TRUE, rarify_n = 4, rarify_nit = 5, nloci = nrow(subvcf@gt))
par(mfrow = c(1,3))
plot(nr[[1]], col = magma(100), zlim = c(1.170413, 1.999993))
plot(yr[[1]], col = magma(100), zlim = c(1.170413, 1.999993))
plot(mask(full_ar, nr)[[1]], col = magma(100), zlim = c(1.170413, 1.999993))

par(mfrow = c(1,3))
plot(nr[[1]], col = magma(100))
plot(yr[[1]], col = magma(100))
plot(mask(full_ar, nr)[[1]], col = magma(100))
```


# check grid alignment
```{r, fig.width = 10, fig.height = 20}
library(wingen)
library(raster)
library("viridis")
data("lotr_lyr")

lyr <- init(lotr_lyr, 'chess')
lyr3 <- init(aggregate(lyr, 3),  'chess')
lyrd3 <- init(disaggregate(lyr, 3),  'chess')
pts <- raster::rasterToPoints(lyr)
pts3 <- raster::rasterToPoints(lyr3)

par(pty = "s", mfrow = c(4,2))
plot(lyr, col = mako(2, begin = 0.3, end = 0.5), box = FALSE, axes = FALSE, legend = FALSE)
points(pts3, pch = 3, col = magma(1, begin = 0.6), cex = 0.5)
plot(lyr, col = mako(2, begin = 0.4, end = 0.6), xlim = c(0,10), ylim = c(-10,0), box = FALSE, axes = FALSE, legend = FALSE)
points(pts3[,c("x","y")], pch = 3, col = magma(1, begin = 0.75), cex = 4, lwd = 3, xlim = c(0,10), ylim = c(-10,0))

plot(lyr3, col = mako(2, begin = 0.3, end = 0.5), box = FALSE, axes = FALSE, legend = FALSE)
points(pts3, pch = 3, col = magma(1, begin = 0.6), cex = 0.5)
plot(lyr3, col = mako(2, begin = 0.4, end = 0.6), xlim = c(0,10), ylim = c(-10,0), box = FALSE, axes = FALSE, legend = FALSE)
points(pts3[,c("x","y")], pch = 3, col = magma(1, begin = 0.75), cex = 4, lwd = 3, xlim = c(0,10), ylim = c(-10,0))

plot(lyr, col = mako(2, begin = 0.3, end = 0.5), box = FALSE, axes = FALSE, legend = FALSE)
points(pts, pch = 3, col = magma(1, begin = 0.6), cex = 0.1)
plot(lyr, col = mako(2, begin = 0.4, end = 0.6), xlim = c(0,10), ylim = c(-10,0), box = FALSE, axes = FALSE, legend = FALSE)
points(pts[,c("x","y")], pch = 3, col = magma(1, begin = 0.75), cex = 4, lwd = 3, xlim = c(0,10), ylim = c(-10,0))

plot(lyrd3, col = mako(2, begin = 0.3, end = 0.5), box = FALSE, axes = FALSE, legend = FALSE)
points(pts, pch = 3, col = magma(1, begin = 0.6), cex = 0.1)
plot(lyrd3, col = mako(2, begin = 0.4, end = 0.6), xlim = c(0,10), ylim = c(-10,0), box = FALSE, axes = FALSE, legend = FALSE)
points(pts[,c("x","y")], pch = 3, col = magma(1, begin = 0.75), cex = 1, lwd = 3, xlim = c(0,10), ylim = c(-10,0))



```
# Heterozygosity is weird
```{r, fig.width = 10, fig.height = 5}
#hetmat <- vcfR::is.het(vcfR::extract.gt(vcf), na_is_false = FALSE)

#write.csv(hetmat, "hetmat_full.csv")
hetmat <- read.csv("hetmat_full.csv")
hetmat <- hetmat[,-1]
hetmat <- as.matrix(hetmat)

rM <- rowMeans(hetmat)
hist(rM)
cM <- colMeans(hetmat)
hist(cM)
M <- mean(hetmat)

rM100 <- rowMeans(hetmat[, 1:100], na.rm = TRUE)
cM100 <- colMeans(hetmat[, 1:100], na.rm = TRUE)
M100 <- mean(hetmat[, 1:100], na.rm = TRUE)

rM200 <- rowMeans(hetmat[, 1:200], na.rm = TRUE)
cM200 <- colMeans(hetmat[, 1:200], na.rm = TRUE)
M200 <- mean(hetmat[, 1:200], na.rm = TRUE)

par(mfrow = c(1,2), pty = "s")
plot(1, type = "n", xlim = c(0, 0.7), ylim = c(0,15), ylab = "density", xlab = "H", main = "Loci Averages (n = 150k)")
polygon(density(rM), col = rgb(1,1,0,0.5))
polygon(density(rM100), col = rgb(0,1,1,0.5))
polygon(density(rM200), col = rgb(1,0,1,0.5))

legend("topright",c("full", "100 samples", "200 samples"), col = c("yellow", "cyan", "magenta"), pch = 15)


plot(1, type = "n", xlim = c(0, 0.7), ylim = c(0,15), ylab = "density", xlab = "H", main = "Individual Averages (n = 1k)")
polygon(density(cM), col = rgb(1,1,0,0.5))
polygon(density(cM100), col = rgb(0,1,1,0.5))
polygon(density(cM200), col = rgb(1,0,1,0.5))

legend("topright",c("full", "100 samples", "200 samples"), col = c("yellow", "cyan", "magenta"), pch = 15)

mean(cM)
mean(rM)
M
mean(cM100)
mean(rM100)
M100
mean(cM200)
mean(rM200)
M200


```

```{r, fig.width = 15, fig.height=3}
msk_lyr <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = 100)


par(mfrow = c(1,5))
for(i in c(100, 200, 300, 500)){
  r <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = i)
  r <- mask(r, msk_lyr)
  plot(r,  col = magma(100), box = FALSE, axes = FALSE, main = paste(i), zlim = c(0.3,0.5))    
}


msk_lyr <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = 200)

par(mfrow = c(1,5))
for(i in c(200, 300, 500)){
  r <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = i)
  r <- mask(r, msk_lyr)
  plot(r,  col = magma(100), box = FALSE, axes = FALSE, main = paste(i), zlim = c(0.3,0.5))    
}
```

```{r, fig.width = 5, fig.height = 5}
plot(1, type = "n", xlim = c(0, 0.7), ylim = c(0,15), ylab = "density", xlab = "H", main = "Individual Averages (n = 1k)")
polygon(density(cM), col = rgb(1,1,0,0.5))
polygon(density(cM100), col = rgb(0,1,1,0.5))
polygon(density(cM200), col = rgb(1,0,1,0.5))

r <- get_divout(file.name = "WGS", rarify = TRUE, measure = "heterozygosity", nsamp = "100")
r100 <- na.omit(getValues(r[[1]]))

r <- get_divout(file.name = "WGS", rarify = TRUE, measure = "heterozygosity", nsamp = "200")
r200 <- na.omit(getValues(r[[1]]))

r <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = "300")
r300 <- na.omit(getValues(r[[1]]))

r <- get_divout(file.name = "FULL", rarify = TRUE, measure = "heterozygosity")
rFULL <- na.omit(getValues(r[[1]]))


par(pty = "s")
plot(1, type = "n", xlim = c(0.3, 0.5), ylim = c(0,45), ylab = "density", xlab = "H", main = "Raster Values")
polygon(density(rFULL), col = rgb(1,1,0,0.5))
polygon(density(r100), col = rgb(0,1,1,0.5))
polygon(density(r200), col = rgb(1,0,1,0.5))
legend("topright",c("full", "100 samples", "200 samples"), col = c("yellow", "cyan", "magenta"), pch = 15)

```

```{r, fig.height = 5, fig.width = 5}
par(mfrow = c(2,2), mar = rep(2,4))

for(nsamp in c(100, 200)){
    rs <- get_divout(file.name = "WGS", rarify = "TRUE", measure = "heterozygosity", nsamp = nsamp)
    plot(rs,  col = magma(100), box = FALSE, axes = FALSE)          
  
    r <- get_divout(file.name = "FULL", rarify = "TRUE", measure = "heterozygosity")
    r <- mask(r, rs)
    plot(r,  col = magma(100), box = FALSE, axes = FALSE) 
}

mr <- get_divout(file.name = "WGS", rarify = "TRUE", measure = "heterozygosity", nsamp = 100)
for(nsamp in c(100, 200)){
    rs <- get_divout(file.name = "WGS", rarify = "TRUE", measure = "heterozygosity", nsamp = nsamp)
    rs <- mask(rs, mr)
    plot(rs,  col = magma(100), box = FALSE, axes = FALSE, zlim =  c(0.3,0.5))          
  
    r <- get_divout(file.name = "FULL", rarify = "TRUE", measure = "heterozygosity")
    r <- mask(r, mr)
    plot(r,  col = magma(100), box = FALSE, axes = FALSE, zlim = c(0.3,0.5)) 
}
```


```{r, fig.width = 5, fig.height = 5}

r1 <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = "100")
r100 <- na.omit(getValues(r1[[1]]))

r3 <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = "300")
r300 <- na.omit(getValues(r3[[1]]))

r5 <- get_divout(file.name = "vn", rarify = TRUE, measure = "heterozygosity", nsamp = "500")
r500 <- na.omit(getValues(r5[[1]]))

rF <- get_divout(file.name = "FULL", rarify = TRUE, measure = "heterozygosity")
rFULL <- na.omit(getValues(rF[[1]]))



cp <- viridis::plasma(4, 0.5)
par(pty = "s")
plot(1, type = "n", xlim = c(0.3, 0.5), ylim = c(0,45), ylab = "density", xlab = "H", main = "Raster Values")

polygon(density(rFULL), col = cp[4])
polygon(density(r500), col = cp[3])
polygon(density(r300), col = cp[2])
polygon(density(r100), col = cp[1])
legend("topright",c("100", "300", "500", "FULL"), col = cp, pch = 15)

```
```{r, fig.height = 8, fig.width = 10}
par(pty = "s", mfrow = c(3,4), mar = rep(1,4))
plot_gd(r1)
plot_gd(r3)
plot_gd(r5)
plot_gd(rF)

plot_gd(r1)
plot_gd(mask(r3, r1))
plot_gd(mask(r5, r1))
plot_gd(mask(rF, r1))

plot_gd(r1, zlim = c(0.25, 0.45))
plot_gd(mask(r3, r1), zlim = c(0.25, 0.45))
plot_gd(mask(r5, r1), zlim = c(0.25, 0.45))
plot_gd(mask(rF, r1), zlim = c(0.25, 0.45))
```
# NEW SIMS
```{r}
hf <- raster("outputs/FULL_rarifyTRUE_nsamp1450_nloci150000_heterozygosity.tif")
plot(hf, col = magma(100))
```

